Fix styles declarations returning before all properties output

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #42913.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

When `useRootPaddingAwareAlignments` is set, `getStylesDeclarations` returns before all the global style properties are output. This means colour changes in global styles aren't reflected in the editor.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Change the function logic so it only omits `padding` properties when `useRootPaddingAwareAlignments` is enabled, as well as _not_ outputting the variables if the setting is disabled.

Adds tests to check for correct behaviour.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. In theme.json settings, add `"useRootPaddingAwareAlignments": true`;
2. In the global styles sidebar in the site editor, change the color or background settings;
3. Check that those changes have a visible effect in the editor;
4. Change `"useRootPaddingAwareAlignments"` back to `false`;
5. Check that everything still works as expected.
6. As a bonus, check that either the variables _or_ the padding properties are output for `.editor-styles-wrapper`, but not both.

## Screenshots or screencast <!-- if applicable -->

tellthemachines committed:
Fix styles declarations returning before all properties output

andrewserong commented on packages/edit-site/src/components/global-styles/use-global-styles-output.js:
```diff
@@ -211,6 +211,16 @@ function getStylesDeclarations(
+			if (
+				( key === '--wp--style--root--padding' &&
+					typeof styleValue === 'string' ) ||
+				! useRootPaddingAlign
```
For this condition, is it possible that more declarations will be skipped than desired if `! useRootPaddingAlign` evaluates to `true`? I was wondering if the condition is meant to be just for the `--wp--style--root--padding` key, like:

```js
			if (
				key === '--wp--style--root--padding' &&
				( typeof styleValue === 'string' || ! useRootPaddingAlign )
			) {
				return declarations;
			}
```

Apologies if I'm not understanding the logic here correctly!

andrewserong replied:
It looks like `fontFamily` is one of the few ones that isn't using the style engine, so I found it a little challenging to work out an error state for this one, but I _think_ the following test seems to catch it:

```js
		it( 'Should output padding and font family properties if useRootPaddingAwareAlignments is disabled', () => {
			const blockStylesWithFontFamily = {
				...blockStyles,
				typography: {
					fontFamily: 'sans-serif',
				},
			};
			expect(
				getStylesDeclarations(
					blockStylesWithFontFamily,
					'body',
					false
				)
			).toEqual( [
				'font-family: sans-serif',
				'background-color: var(--wp--preset--color--light-green-cyan)',
				'padding-top: 33px',
				'padding-right: 33px',
				'padding-bottom: 33px',
				'padding-left: 33px',
			] );
		} );
```

andrewserong replied:
Oh, I've found a better real world example with the Duotone feature. In TwentyTwentyTwo, if you add the following to the `theme.json` file (from https://github.com/WordPress/gutenberg/pull/37727) to switch on a particular Duotone for image blocks:

```json
{
	"styles": {
		"blocks": {
			"core/image": {
				"filter": {
					"duotone": "var(--wp--preset--duotone--foreground-and-secondary)"
				}
			}
		}
	}
}
```

| Expected | Actual |
| --- | --- |
| <img width="1275" alt="image" src="https://user-images.githubusercontent.com/14988353/182771891-6c06f11c-e948-42d5-8933-87023a930d2a.png"> | <img width="1275" alt="image" src="https://user-images.githubusercontent.com/14988353/182772063-3ef0105b-3525-46ac-a66c-244dcf02fae7.png"> |

The above change appears to fix it.

tellthemachines replied:
Oh, good catch! Yes, the parentheses should be around the OR as you suggested 😄

andrewserong reviewed (commented):
Nice work figuring out this one @tellthemachines (and for adding additional tests 🎉 )! Skipping the output in the loop over `extraRules` looks like a good approach to me 👍. I just left a comment about the condition to skip the variable output, as it looks like it might be unexpectedly returning early.

Aside from that though, this is otherwise testing nicely for me manually in the site editor with TwentyTwentyTwo and Emptytheme and resolves the issue with color output.

tellthemachines committed:
Fix conditional bug and improve tests.

tellthemachines commented:
@andrewserong thanks for reviewing and kindly providing the extra test! I fixed the issue, and added font-family to the initial block styles in the test, so now we can check that it gets output on all three tests.

andrewserong reviewed (approved):
Excellent, thanks for fixing that up, this is working nicely for me now @tellthemachines! 🎉 

<img width="400" alt="image" src="https://user-images.githubusercontent.com/14988353/182780311-9f4a2f66-cd7c-40ab-bee0-f21ccc31a186.png">

LGTM 🚀

ndiego commented:
@tellthemachines, @andrewserong, @andrewserong would this be a candidate for a 13.8.x release? The root issue is impacting quite a range of things. I just discovered that some theme.json style settings are not getting output, notably those for template parts. With this fix, everything now works like a charm.

tellthemachines commented:
@ndiego it's not a huge change so should be fine to add to the next minor release, yes!
